### DISCUSSION TITLE
Pagination function?

### DISCUSSION THREAD

**Discussion by 3l3ktr0n** [e107inc/e107#4642] [Q&A Plugins] [answered]
Hey! Sorry it's me again with my stupid questions. Does e107 have a function for easy pagination by itself or do I have to squeeze it out of the news plugin somehow?

**Comment by Moc** [comment]
It is currently not documented yet (work in progress!), but there is a public function for it: `e_form::pagination()`. 

You can call it by using `e107::getForm()->pagination()`. The method can be found in `e107_handlers/form_handler.php`:
```php
        /**
	 * Render Pagination using 'nextprev' shortcode.
	 * @param string $url eg. e_REQUEST_SELF.'?from=[FROM]'
	 * @param int $total total records
	 * @param int $from value to replace [FROM] with in the URL
	 * @param int $perPage number of items per page
	 * @param array $options template, type, glyphs
	 * @return string
	*/
	public function pagination($url='', $total=0, $from=0, $perPage=10, $options=array())
	{
	...
	}
```

Alternatively, you can use the {NEXTPREV} shortcode directly. You can find more information on it in `e107_core/shortcodes/single/nextprev.php`.

**Reply by 3l3ktr0n** [reply]
okay, I got it to parse the pagination itself... how can I make with work with my actual records now?

**Reply by Moc** [reply]
@3l3ktr0n I wrote an example code with items from the chatbox plugin (as I happened to have many records stored in the chatbox). 

```php
<?php

require_once("class2.php");

require_once(HEADERF);

$text = "";

$total 		= e107::getDb()->count("chatbox"); 
$from 		= (int) $_GET['from']; 
$perPage 	= 5; 

$items = e107::getDb()->retrieve("chatbox", "cb_id, cb_nick, cb_message","LIMIT ".$from.", ".$perPage."", true);


foreach($items as $item)
{
	$text .= $item['cb_id'] ." - ".$item['cb_nick'].": ".$item['cb_message']."<br>";
}


$text .= e107::getForm()->pagination(e_REQUEST_SELF.'?from=[FROM]', $total, $from, $perPage); 

e107::getRender()->tablerender("Pagination Test", $text); 

require_once(FOOTERF);
```

**Comment by 3l3ktr0n** [comment, marked as answer]
This pagination example is for displaying data per pages not as records (page vs from).

```php
<?php

include_once('class2.php');

require_once(HEADERF);

$frm = e107::getForm();
$sql = e107::getDb();
$ns = e107::getRender();

$text = "";
$count = $sql->count('chatbox');
$page = isset($_GET['page']) ? $_GET['page'] : 1;
$perPage = 10;
$from = ($page - 1) * $perPage;
$total = ceil($count / $perPage); 
$options = array('type' => 'page', 'navcount' => 4);

$rows = $sql->retrieve("chatbox", "cb_id, cb_nick, cb_message","LIMIT ".$from.", ".$perPage."", true);

foreach($items as $item)
{
	$text .= $item['cb_id'] ." - ".$item['cb_nick'].": ".$item['cb_message']."<br>";
}

$text = $frm->pagination(e_REQUEST_SELF.'?page=[FROM]', $total, $page, $perPage, $options); 

$ns->tablerender("Pagination Test", $text); 

require_once(FOOTERF);

?>
```

**Reply by Jimmi08** [reply]
Thank you so much for this. I spend hours to get the correct **page** pagination.
This was the trick:
```
$total = ceil($count / $perPage); 
```
But I consider this as a bug.  The pagination code should manage this itself. If my type is "page", then nexprev shortcode should change total value as you did.  
But your code example is very good example.
